bucket.go

```diff
@@ -144,9 +144,55 @@ func (b *Bucket) openBucket(value []byte) *Bucket {
 		child.page = (*page)(unsafe.Pointer(&value[bucketHeaderSize]))
 	}
 
-	return &child
+	return child
+}
+
+// Close releases the bucket structure so the memory can be used for another bucket.
+// Do not touch the bucket again after closing it. Note this does not effect the
+// contents of the bucket in the database, just the memory allocated to control
+// access to it.
+// You do not need to call Close() on your buckets, but doing so may reduce the
+// number of memory allocations, particularly when traversing a large number of
+// buckets in a read transaction.
+func (b *Bucket) Close() {
+	// if the bucket is part of a writeable transaction then we don't re-use the
+	// memory as it is kept in its parent buckets map
+	if b.tx.writable {
+		return
+	}
+	bucketPool.Put(b)
+}
+
+type freeBuckets struct {
+	sync.Mutex
+	items []*Bucket
 }
 
+func (fs *freeBuckets) Get() *Bucket {
+	fs.Lock()
+	l := len(fs.items)
+	var i *Bucket
+	if l > 0 {
+		i = fs.items[l-1]
+		fs.items = fs.items[:l-1]
+	} else {
+		i = &Bucket{}
+	}
+	fs.Unlock()
+	return i
+}
+
+func (fs *freeBuckets) Put(i *Bucket) {
+	// Ensure bucket is cleaned before storage - otherwise we might stop things
+	// being cleaned up with GC
+	*i = Bucket{}
+	fs.Lock()
+	fs.items = append(fs.items, i)
+	fs.Unlock()
+}
+
+var bucketPool = freeBuckets{}
```

Comment:
Is there are reason this doesn't use a sync.Pool? sync.Pool allows the values to still be gc'ed and could even be faster (I haven't testet this). Also sync.Pool has some special optimizations, so it will probably be faster when using multiple goroutines concurrently.